man/man1/lgt2txt.1
.TH lgt2txt 1 "June 10, 2010" "Logtalk 2.44.1" "Logtalk Documentation"

.SH NAME
lgt2txt \-- Logtalk script for converting XML documenting files into text files

.SH SYNOPSIS
.B lgt2txt
[
.BI \-d\ directory
]
[
.BI \-p\ processor
]
.br
.B lgt2txt
[
.BI \-h
]

.SH DESCRIPTION
\f3lgt2txt\f1 is a Logtalk script for converting all Logtalk XML documenting files in the current directory into text files. Assumes UTF-8 encoding.

.SH OPTIONS
.TP
.BI \-d\ directory
Output directory for the text files.
.TP
.BI \-p\ processor
XSLT processor.
.TP
.BI \-h
Prints a summary of the script options.

.SH REQUIREMENTS
Requires a compatible XSLT processor. See the script help screen for more information.

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH FILES
.TP
.BI $LOGTALKHOME/xml/logtalk.dtd
The Logtalk DTD file.
.TP
.BI $LOGTALKUSER/xml/custom.ent
The Logtalk custom entities file.
.TP
.BI $LOGTALKHOME/xml/logtalk.xsd
The Logtalk XML Schema file.
.TP
.BI $LOGTALKUSER/xml/lgttxt.xsl
The default XSLT file.

.SH "SEE ALSO"
The Logtalk website at http://logtalk.org/
.PP
The Logtalk User and Reference Manuals at $LOGTALKHOME/manuals/index.html
.PP
logtalk_user_setup(1),\ lgt2html(1),\ lgt2xml(1),\ lgt2pdf(1)

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2012 Paulo Moura.


RELEASE_NOTES
=================================================================
Logtalk - Object oriented extension to Prolog
Release 2.28.0

Copyright (c) 1998-2006 Paulo Moura.  All Rights Reserved.
=================================================================


RELEASE NOTES
=============


2.28.0 - June ??, 2006

	Added support for multi-threading programming on selected Prolog compilers.
	Added a new object directive, threaded/0, and four new built-in predicates,
	threaded_call/1-2 and threaded_exit/1-2. Added a new set of exemples, 
	"threads".

	Updated the config file for the forthcoming SICStus Prolog 4.0 version 
	to work in the current beta version. Modified the "make_sicstus.lgt.js" 
	script to work with both SICStus Prolog 3.12 and 4.0 versions.

	Updated the config file for YAP to workaround a problem on the Windows 
	version where testing for directory existence fails if the path ends with 
	a slash (or a double backslash); this problem prevents the use of library 
	notation to load source files.

	Dropped support for versions of XSB older than 3.0.

	Extended "parametric" and "metainterpreters" examples.

	Update all the JScript scripts to check if a compatible version of WSH is 
	installed and to workaround some problems with spaces in file paths.

	Added Inno Setup GUI Windows installer script.


2.27.1 - March 27, 2006

	Allow calls to the built-in method parameter/2 with the first argument 
	not instantiated at compile time. Detect and report as a compilation 
	error parameter/2 indexes that are out of range.

	Optimized generation of predicate definition and declaration linking 
	clauses, resulting in small lookup performance improvements and in space 
	savings for the Prolog code generated when compiling Logtalk entities.

	Many minor code and documentation improvements to the Logtalk compiler.

	Added an object version of the tokenizer described in the Michael 
	Covington's paper "Tokenization using DCG Rules" to the "dcgs" example.

	Improved integration code for Qu-Prolog and Logtalk ("configs/qphook.ql").

	Improved library hierarchy methods that return lists of objects in order 
	to avoid duplicated elements (library objects "proto_hierarchy.lgt" and 
	"class_hierarchy.lgt"). Added new methods extension/1 and extensions/1 to 
	the library object "proto_hierarchy.lgt".

	Documented the concept of "parametric object proxy". Added a new example, 
	"proxies", illustrating the use of parametric object proxies.

	Added support for code completion and for listing entity names on the 
	symbol pop-up menu to the MacOS X TextMate text editor.

	Updated the "cplgtdirs.*" scripts to also create an alias/shortcut to the 
	"wenv" directory.

	Renamed the alternative compilation and documentation directory names on 
	all config files to be compatible across operating-systems. Removed unused
	predicate '$lgt_reverse'/2 from all config files.

	Updated HTML manuals "print.css" CSS file for compatibility with the 
	latest version of CSSToXSLFO.


2.27.0 - February 9, 2006

	Improved performance for local calls to the built-in methods phrase/2-3.
	Allow the built-in methods phrase/2-3 to accept both partial and proper 
	lists of terminals.

	Improved grammar rule translator. Report calls to undefined non-terminals 
	when translating grammar rules.

	Added support for declaring grammar rule non-terminal aliases using the 
	alias/3 directive.

	Added a new predicate property, non_terminal/1, for predicates resulting 
	from the compilation of grammar rule non-terminals.

	Improved support for the built-in method expand_term/2 in order to allow 
	bypassing of the default Logtalk grammar rule translator by defining 
	clauses for the term_expansion/2 predicate. Added a "term_expansionp" 
	protocol to the Logtalk library. Added a new example, "expansion", 
	illustrating the use of the term_expansion/2 predicate.

	Added a new compiler flag, hook/1, allowing the specification of a 
	compiler hook that is called for which term read from a source file.
	Added a simple example, "hooks", of using the Logtalk compiler hook 
	to expand author abbreviations in info/1 directives into full names 
	and email addresses.

	Added support for XSB 3.x to the runtime error handler. Updated the XSB 
	3.x config file with declarations for multi-threading meta-predicates.
	
	Removed a few choice-points when checking validity of entity directives.

	Added two new objects, list(Type) and set(Type), to the standard library 
	supporting the validation of lists and ordered sets whose elements are 
	restricted to a single type.

	Added a new DCG example for solving enigmas encoded using a cellphone 
	keypad.

	Added a missing library dependency to the "puzzles" example SCRIPT file.

	Removed the experimental "systemp.lgt" protocol from the list of files 
	loaded by the "library/all_loader.lgt" loader utility files to avoid 
	compilation errors on some Prolog systems.

	Corrected a bug that prevented dynamic creation of categories using the 
	built-in predicate create_category/4.

	Corrected a bug in the reporting of singleton variables, which failed to 
	write an accurate message for facts and grammar rules.

	Corrected a bug in passing the correct calling context ("this") when 
	processing meta-calls in objects.

	Corrected a bug in scope checking with local calls to reflection and 
	database methods.

	Corrected a bug in checking the validity of the arguments of the op/3 
	directive appearing inside entities.

	Added predicates for testing if a term is a partial list or a proper list 
	to all config files.

	Added a definition for missing open/4 ISO Prolog predicate to the config 
	file of Bin Prolog.

	Added a workaround for a B-Prolog bug to this compiler config file.


2.26.2 - December 20, 2005

	Improved error-checking for the Logtalk compiler and for the grammar rule 
	translator.

	Small performance improvements for message sending and for the built-in 
	database methods.


xml/lgt2txt.js
// ================================================================
// Logtalk - Open source object-oriented logic programming language
// Release 2.44.1
//
// Copyright (c) 1998-2012 Paulo Moura. All Rights Reserved.
// 
// Logtalk is free software. You can redistribute it and/or modify
// it under the terms of the Artistic License 2.0 as published by 
// the The Perl Foundation.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// Artistic License 2.0 for more details. A copy of the license is 
// provided in the "LICENSE.txt" file.
// ================================================================

if (ScriptEngineMajorVersion() < 5 || ScriptEngineMajorVersion() == 5 && ScriptEngineMinorVersion() < 6) {
	WScript.Echo('Error! WSH 5.6 or later version needed for running this script.');
	WScript.Quit(1);
}

var WshShell = new ActiveXObject("WScript.Shell");

var directory = WshShell.CurrentDirectory;

var processor = "msxsl";
// var processor = "xsltproc";
// var processor = "xalan";
// var processor = "sabcmd";

if (WScript.Arguments.Unnamed.Length > 0) {
	usage_help();
	WScript.Quit(0);
}

var WshSystemEnv = WshShell.Environment("SYSTEM");
var WshUserEnv = WshShell.Environment("USER");

var logtalk_home;
var logtalk_user;

if (WshSystemEnv.Item("LOGTALKHOME"))
	logtalk_home = WshSystemEnv.Item("LOGTALKHOME");
else if (WshUserEnv.Item("LOGTALKHOME"))
	logtalk_home = WshUserEnv.Item("LOGTALKHOME")
else {
	WScript.Echo("Error! The environment variable LOGTALKHOME must be defined first!");
	usage_help();
	WScript.Quit(1);
}

if (WshSystemEnv.Item("LOGTALKUSER"))
	logtalk_user = WshSystemEnv.Item("LOGTALKUSER");
else if (WshUserEnv.Item("LOGTALKUSER"))
	logtalk_user = WshUserEnv.Item("LOGTALKUSER")
else {
	WScript.Echo("Error! The environment variable LOGTALKUSER must be defined first!");
	usage_help();
	WScript.Quit(1);
}

var xslt = logtalk_user + "\\xml\\lgttxt.xsl";

var d_arg = "";
var p_arg = "";

if (WScript.Arguments.Named.Exists("d"))
	d_arg = WScript.Arguments.Named.Item("d");

if (WScript.Arguments.Named.Exists("p"))
	p_arg = WScript.Arguments.Named.Item("p");

var FSObject = new ActiveXObject("Scripting.FileSystemObject");

if (d_arg != "" && !FSObject.FolderExists(d_arg)) {
	WScript.Echo("Error! directory does not exists: " + d_arg);
	WScript.Echo("");
	usage_help();
} else if (d_arg != "")
	directory = d_arg;

if (p_arg != "" && p_arg != "msxsl" && p_arg != "xsltproc" && p_arg != "xalan" && p_arg != "sabcmd") {
	WScript.Echo("Error! Unsupported XSLT processor:" + p_arg);
	WScript.Echo("");
	usage_help();
} else if (p_arg != "")
	processor = p_arg;

if (!FSObject.FileExists(WshShell.CurrentDirectory + "\\logtalk.dtd")) {
	FSObject.CopyFile(logtalk_home + "\\xml\\logtalk.dtd", WshShell.CurrentDirectory + "\\logtalk.dtd");
}

if (!FSObject.FileExists(WshShell.CurrentDirectory + "\\custom.ent")) {
	FSObject.CopyFile(logtalk_home + "\\xml\\custom.ent", WshShell.CurrentDirectory + "\\custom.ent");
}

if (!FSObject.FileExists(WshShell.CurrentDirectory + "\\logtalk.xsd")) {
	FSObject.CopyFile(logtalk_home + "\\xml\\logtalk.xsd", WshShell.CurrentDirectory + "\\logtalk.xsd");
}

if (!FSObject.FileExists(directory + "\\logtalk.css")) {
	FSObject.CopyFile(logtalk_user + "\\xml\\logtalk.css", directory + "\\logtalk.css");
}

WScript.Echo("");
WScript.Echo("converting XML files...");

var files = new Enumerator(FSObject.GetFolder(WshShell.CurrentDirectory).Files);

for (files.moveFirst(); !files.atEnd(); files.moveNext()) {
	var file = files.item().name;
	if (FSObject.GetExtensionName(file) == "xml") {
		WScript.Echo("  converting " + file);
		var txt_file = directory + "\\" + FSObject.GetBaseName(file) + ".txt";
		switch (processor) {
			case "msxsl" :
				WshShell.Run("msxsl -o \"" + txt_file + "\" \"" + file + "\" \"" + xslt + "\"", true);
				break;
			case "xsltproc" :
				WshShell.Run("xsltproc -o \"" + txt_file + "\" \"" + xslt + "\" \"" + file + "\"", true);
				break;
			case "xalan" :
				WshShell.Run("xalan -o \"" + txt_file + "\" \"" + file + "\" \"" + xslt + "\"", true);
				break;
			case "sabcmd" :
				WshShell.Run("sabcmd \"" + xslt + "\" \"" + file + "\" \"" + txt_file + "\"", true);
				break;
		}
	}
}

WScript.Echo("conversion done");
WScript.Echo("");

WScript.Quit(0);

function usage_help() {
	WScript.Echo("");
	WScript.Echo("This script converts all Logtalk XML files documenting files in the");
	WScript.Echo("current directory to text files");
	WScript.Echo("");
	WScript.Echo("Usage:");
	WScript.Echo("  " + WScript.ScriptName + " [/d:directory] [/p:processor]");
	WScript.Echo("  " + WScript.ScriptName + " help");
	WScript.Echo("");
	WScript.Echo("Optional arguments:");
	WScript.Echo("  d - output directory for the generated files (default is " + directory + ")");
	WScript.Echo("  p - XSLT processor (msxsl, xsltproc, xalan, or sabcmd; default is " + processor + ")");
	WScript.Echo("");
	WScript.Quit(1);
}
